# Lab | Random Forests

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split

In [2]:
df1 = pd.read_csv('files_for_lab/categorical.csv')
df2 = pd.read_csv('files_for_lab/numerical.csv')
df3 = pd.read_csv('files_for_lab/target.csv')
data = pd.concat([df1, df2, df3], axis = 1)
data['TARGET_B'].value_counts()

0    90569
1     4843
Name: TARGET_B, dtype: int64

In [3]:
data.shape

(95412, 339)

In [4]:
data.isna().sum()

STATE       0
CLUSTER     0
HOMEOWNR    0
GENDER      0
DATASRCE    0
           ..
HPHONE_D    0
RFA_2F      0
CLUSTER2    0
TARGET_B    0
TARGET_D    0
Length: 339, dtype: int64

In [5]:
data.head()

STATE  CLUSTER HOMEOWNR GENDER  DATASRCE RFA_2R RFA_2A GEOCODE2 DOMAIN_A  \
0    IL       36        H      F         3      L      E        C        T   
1    CA       14        H      M         3      L      G        A        S   
2    NC       43        U      M         3      L      E        C        R   
3    CA       44        U      F         3      L      E        C        R   
4    FL       16        H      F         3      L      F        A        S   

   DOMAIN_B  ODATEW_YR  ODATEW_MM  DOB_YR  DOB_MM  MINRDATE_YR  MINRDATE_MM  \
0         2         89          1      37      12           92            8   
1         1         94          1      52       2           93           10   
2         2         90          1       0       2           91           11   
3         2         87          1      28       1           87           11   
4         2         86          1      20       1           93           10   

   MAXRDATE_YR  MAXRDATE_MM  LASTDATE_YR  LASTDATE_MM  FIRSTDATE_YR  \
0           94            2           95           12            89   
1           95           12           95           12            93   
2           92            7           95           12            90   
3           94           11           95           12            87   
4           96            1           96            1            79   

   FIRSTDATE_MM  TCODE        AGE  INCOME  WEALTH1  HIT  MALEMILI  MALEVET  \
0            11      0  60.000000       5        9    0         0       39   
1            10      1  46.000000       6        9   16         0       15   
2             1      1  61.611649       3        1    2         0       20   
3             2      0  70.000000       1        4    2         0       23   
4             3      0  78.000000       3        2   60         1       28   

   VIETVETS  WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  WEALTH2  POP901  POP902  \
0        34        18        10         2       1        5     992     264   
1        55        11         6         2       1        9    3611     940   
2        29        33         6         8       1        1    7001    2040   
3        14        31         3         0       3        0     640     160   
4         9        53        26         3       2        9    2520     627   

   POP903  POP90C1  POP90C2  POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  \
0     332        0       35       65       47       53    92     1     0   
1     998       99        0        0       50       50    67     0     0   
2    2669        0        2       98       49       51    96     2     0   
3     219        0        8       92       54       46    61     0     0   
4     761       99        0        0       46       54     2    98     0   

   ETH4  ETH5  ETH6  ETH7  ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  \
0     0    11     0     0     0     0      0      0      0     11      0   
1    31     6     4     2     6     4     14      0      0      2      0   
2     0     2     0     0     0     0      0      0      0      2      0   
3    11    32     6     2     0     0      0      0      0     31      0   
4     0     1     0     0     0     0      0      0      0      0      0   

   ETH15  ETH16  AGE901  AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  \
0      0      0      39      48      51      40      50      54      25   
1      1      4      34      41      43      32      42      45      32   
2      0      0      35      43      46      37      45      49      23   
3      0      1      32      40      44      34      43      47      25   
4      0      0      33      45      50      36      46      50      27   

   CHIL1  CHIL2  CHIL3  AGEC1  AGEC2  AGEC3  AGEC4  AGEC5  AGEC6  AGEC7  \
0     31     42     27     11     14     18     17     13     11     15   
1     33     46     21     13     14     33     23     10      4      2   
2     35     40     25     13     20     19     16     13     10      8   
3     45     35     20     1

In [6]:
data['TARGET_B'].value_counts()

0    90569
1     4843
Name: TARGET_B, dtype: int64

### Apply the Random Forest algorithm to predict the TARGET_B. Please note that this column suffers from class imbalance. Fix the class imbalance using upsampling.

In [7]:
# upsampling the minority 
from sklearn.utils import resample
df_majority = data[data['TARGET_B'] == 0]
df_minority = data[data['TARGET_B'] == 1]

df_minority_upsampled = resample(df_minority, 
                                 replace=True,     # sample with replacement
                                 n_samples=len(df_majority),    # to match majority class
                                 random_state=123) # reproducible results
data1 = pd.concat([df_majority, df_minority_upsampled], axis=0)
data1.shape

(181138, 339)

In [8]:
y = data1[['TARGET_B','TARGET_D']]
X = data1.drop(['TARGET_B', 'TARGET_D'], axis = 1)
X.shape

(181138, 337)

In [9]:
high_cardinality_columns = []
for column in X.columns:
    unique_values = X[column].nunique()
    if unique_values > 10:  # You can adjust the threshold (e.g., 10) based on your preference
        high_cardinality_columns.append((column, unique_values))
if high_cardinality_columns:
    print("Columns with high cardinality:")
    for column, cardinality in high_cardinality_columns:
        print(f"{column}: {cardinality} unique values")
else:
    print("No columns with high cardinality found.")
    
X1 = X.loc[:, ~X.columns.isin(high_cardinality_columns)]

Columns with high cardinality:
STATE: 12 unique values
CLUSTER: 53 unique values
ODATEW_YR: 15 unique values
ODATEW_MM: 12 unique values
DOB_YR: 96 unique values
DOB_MM: 12 unique values
MINRDATE_YR: 20 unique values
MINRDATE_MM: 12 unique values
MAXRDATE_YR: 18 unique values
MAXRDATE_MM: 12 unique values
LASTDATE_MM: 12 unique values
FIRSTDATE_YR: 26 unique values
FIRSTDATE_MM: 12 unique values
TCODE: 55 unique values
AGE: 97 unique values
HIT: 75 unique values
MALEMILI: 95 unique values
MALEVET: 89 unique values
VIETVETS: 95 unique values
WWIIVETS: 100 unique values
LOCALGOV: 55 unique values
STATEGOV: 65 unique values
FEDGOV: 55 unique values
POP901: 9906 unique values
POP902: 4786 unique values
POP903: 5698 unique values
POP90C1: 100 unique values
POP90C2: 100 unique values
POP90C3: 100 unique values
POP90C4: 81 unique values
POP90C5: 81 unique values
ETH1: 100 unique values
ETH2: 100 unique values
ETH3: 85 unique values
ETH4: 96 unique values
ETH5: 100 unique values
ETH6: 21 uniqu

In [10]:
X1

STATE  CLUSTER HOMEOWNR GENDER  DATASRCE RFA_2R RFA_2A GEOCODE2  \
0         IL       36        H      F         3      L      E        C   
1         CA       14        H      M         3      L      G        A   
2         NC       43        U      M         3      L      E        C   
3         CA       44        U      F         3      L      E        C   
4         FL       16        H      F         3      L      F        A   
...      ...      ...      ...    ...       ...    ...    ...      ...   
16890  other       11        H      F         2      L      F        A   
60980     CA        1        H      F         2      L      G        A   
48158     IL       20        H      F         3      L      E        A   
54769  other       24        H      M         1      L      F        B   
64040     IN       36        U      M         3      L      E        D   

      DOMAIN_A  DOMAIN_B  ODATEW_YR  ODATEW_MM  DOB_YR  DOB_MM  MINRDATE_YR  \
0            T         2         89          1      37      12           92   
1            S         1         94          1      52       2           93   
2            R         2         90          1       0       2           91   
3            R         2         87          1      28       1           87   
4            S         2         86          1      20       1           93   
...        ...       ...        ...        ...     ...     ...          ...   
16890        S         1         95          1      36       1           94   
60980        U         1         92          1      44       1           92   
48158        S         2         86          1       0       2           94   
54769        C         1         86          1      26      12           89   
64040        T         2         86          1      35       6           87   

       MINRDATE_MM  MAXRDATE_YR  MAXRDATE_MM  LASTDATE_YR  LASTDATE_MM  \
0                8           94            2           95           12   
1               10           95           12           95           12   
2               11           92            7           95           12   
3               11           94           11           95           12   
4               10           96            1           96            1   
...            ...          ...          ...          ...          ...   
16890           11           95            7           96            2   
60980            7           95           12           95           12   
48158           12           93            9           96            1   
54769           10           94           12           95           12   
64040           11           95           12           95           12   

       FIRSTDATE_YR  FIRSTDATE_MM  TCODE        AGE  INCOME  WEALTH1  HIT  \
0                89            11      0  60.000000       5        9    0   
1                93            10      1  46.000000       6        9   16   
2                90             1      1  61.611649       3        1    2   
3                87             2      0  70.000000       1        4    2   
4                79             3      0  78.000000       3        2   60   
...             ...           ...    ...        ...     ...      ...  ...   
16890            94            11      2  62.000000       3        9    0   
60980            92             2      0  54.000000       7        9    0   
48158            87             4      2  61.611649       5        9    0   
54769            89            10      0  71.000000       4        8    0   
64040            86            10      1  63.000000       2        2    5   

       MALEMILI  MALEVET  VIETVETS  WWIIVETS  LOCALGOV  STATEGOV  FEDGOV  \
0             0       39        34        18        10         2       1   
1             0       15        55        11         6         2       1   
2             0       20        29        33         6         8       1   
3             0       23        14        31       

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder

numericalX1 = X1.select_dtypes(np.number)
categoricalX1 = X1.select_dtypes(include=['object'])
display(categoricalX1.head())

for col in categoricalX1.columns:
    print(col, categoricalX1[col].nunique())
print()

encoder = OneHotEncoder(drop='first').fit(categoricalX1)
encoded_categorical = encoder.transform(categoricalX1).toarray()
encoded_categorical = pd.DataFrame(encoded_categorical, columns=encoder.get_feature_names_out(), index=categoricalX1.index)



# Concatenate numericalX and encoded_categorical along axis=1
X1 = pd.concat([numericalX1, encoded_categorical], axis=1)

# Note: We need to do train/test split before downsampling, and then only downsample the training set - Why?
X1_train, X1_test, y_train, y_test = train_test_split(X1, y, test_size=0.2, random_state=0)

STATE HOMEOWNR GENDER RFA_2R RFA_2A GEOCODE2 DOMAIN_A
0    IL        H      F      L      E        C        T
1    CA        H      M      L      G        A        S
2    NC        U      M      L      E        C        R
3    CA        U      F      L      E        C        R
4    FL        H      F      L      F        A        S

STATE 12
HOMEOWNR 2
GENDER 3
RFA_2R 1
RFA_2A 4
GEOCODE2 4
DOMAIN_A 5



In [12]:
X1_train.head()

CLUSTER  DATASRCE  DOMAIN_B  ODATEW_YR  ODATEW_MM  DOB_YR  DOB_MM  \
42043       45         2         2         92          1      17       6   
68510        5         2         2         90          1      24       5   
9940         5         3         2         90          1      25       1   
59275        1         3         1         93          1       0       2   
62638       18         3         2         93          1      39       2   

       MINRDATE_YR  MINRDATE_MM  MAXRDATE_YR  MAXRDATE_MM  LASTDATE_YR  \
42043           95            2           95            1           96   
68510           90            4           93           10           95   
9940            94           11           96            9           96   
59275           93           12           95            9           95   
62638           93            2           95            2           95   

       LASTDATE_MM  FIRSTDATE_YR  FIRSTDATE_MM  TCODE        AGE  INCOME  \
42043            3            92            10      1  81.000000       3   
68510           12            90             4     28  74.000000       4   
9940             9            90             1      0  73.000000       5   
59275           11            93             1      1  61.611649       5   
62638            9            93             2      1  59.000000       5   

       WEALTH1  HIT  MALEMILI  MALEVET  VIETVETS  WWIIVETS  LOCALGOV  \
42043        9    0         0       38        24        46        14   
68510        9    0         0       29        46        29        10   
9940         5    2         0       28         8        14         7   
59275        9    0         0       30        11        78         4   
62638        6    0         0       21        40        39         7   

       STATEGOV  FEDGOV  WEALTH2  POP901  POP902  POP903  POP90C1  POP90C2  \
42043         4       4        1    9656    2704    3747        0        0   
68510         2       3        6    1276     332     490       99        0   
9940          7       1        5    1174     297     429       99        0   
59275         0       5        9     764     227     260       99        0   
62638         3       4        6    1587     387     456       99        0   

       POP90C3  POP90C4  POP90C5  ETH1  ETH2  ETH3  ETH4  ETH5  ETH6  ETH7  \
42043       99       50       50    83     0    17     0     1     0     0   
68510        0       49       51    89     0     1     5    22     0     1   
9940         0       47       53    69     3     0    14    26     0     1   
59275        0       50       50    70     0     0    28     5     1     4   
62638        0       50       50    81     1     1    14    15     0     1   

       ETH8  ETH9  ETH10  ETH11  ETH12  ETH13  ETH14  ETH15  ETH16  AGE901  \
42043     0     0      0      0      0      1      0      0      0      36   
68510     1     1      1      0      0     10      1      4      8      37   
9940      1     3      5      1      0     21      1      0      5      36   
59275    22     0      0      0      0      2      0      0      3      44   
62638     1     2      0      8      0     11      0      0      3      30   

       AGE902  AGE903  AGE904  AGE905  AGE906  AGE907  CHIL1  CHIL2  CHIL3  \
42043      47      50      38      49      52      27     39     39     21   
68510      43      45      38      46      49      20     37     37     26   
9940       44      48      38      46      50      21     46     32     23   
59275      49      52      41      50      54      22     23     43     34   
62638      37      40      32      40      44      28     45     36     19   

       AGEC1  AGEC2  AGEC3  AGEC4  AGEC5  AGEC6  AGEC7  CHILC1  CHILC2  \
42043      9     19     19     13     15     13     12      16      17   
68510     12     20     21     16     10     12      9      19      12   
9940      13     20     17     16     15     12      7      18      19   
59275     11      7     18     22     19   

In [13]:
y_train.head()

TARGET_B  TARGET_D
42043         0       0.0
68510         0       0.0
9940          1      30.0
59275         0       0.0
62638         1      20.0

In [14]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix

clf = RandomForestClassifier(max_depth=5,
                             min_samples_split=20,
                             min_samples_leaf =20,
                             max_samples=0.8,
                             random_state = 42)

clf.fit(X1_train, y_train['TARGET_B'])

print("The Accuracy for the Random Forest in the TRAIN set is {:.2f}".format(clf.score(X1_train, y_train['TARGET_B'])))
print("The Accuracy for the Random Forest in the TEST  set is {:.2f}".format(clf.score(X1_test, y_test['TARGET_B'])))

y_pred = clf.predict(X1_test)
display(y_test['TARGET_B'].value_counts())
display(confusion_matrix(y_test['TARGET_B'], y_pred))

The Accuracy for the Random Forest in the TRAIN set is 0.61
The Accuracy for the Random Forest in the TEST  set is 0.61


1    18254
0    17974
Name: TARGET_B, dtype: int64

array([[11163,  6811],
       [ 7281, 10973]])

### Discuss the model predictions and it's impact in the bussiness scenario. Is the cost of a false positive equals to the cost of the false negative? How much the money the company will not earn because of missclassifications made by the model?

In [15]:
y

TARGET_B  TARGET_D
0             0       0.0
1             0       0.0
2             0       0.0
3             0       0.0
4             0       0.0
...         ...       ...
16890         1      21.0
60980         1      20.0
48158         1      10.0
54769         1      15.0
64040         1      12.0

[181138 rows x 2 columns]

When TARGET_B is '1', it means 'cost'. 

Based on the provided results, the Random Forest model achieved an accuracy of approximately 61% both on the training set and the test set. The accuracy indicates the proportion of correctly classified samples out of the total samples. 61% accuracy might seem decent,

True Positives (TP): 11,163

False Positives (FP): 6,811

True Negatives (TN): 10,973

False Negatives (FN): 7,281

The cost of a false positive is slightly less than to the cost of the false negative. This means it is not good news for the comapny. This is because False Negatives (FN) refers to the number of instances where the model incorrectly predicted the negative class as "0" when the true class is "1." The more FN is gotten, the worse for the comapny. 


Total Cost = (Cost of False Positives * Number of False Positives) + (Cost of False Negatives * Number of False Negatives)

In [16]:
y_test['y-pred'] = y_pred

y_test

TARGET_B  TARGET_D  y-pred
66958         0       0.0       1
11231         1       7.0       1
90375         0       0.0       0
47462         0       0.0       0
81846         0       0.0       0
...         ...       ...     ...
55018         0       0.0       0
31519         0       0.0       1
27487         1      50.0       0
26367         1      20.0       1
34643         0       0.0       1

[36228 rows x 3 columns]

In [17]:
FP = y_test[(y_pred == 1) & (y_test['TARGET_B'] == 0)]['TARGET_D'].sum()
FP

0.0

In [18]:
FN = y_test[(y_pred == 0) & (y_test['TARGET_B'] == 1)]['TARGET_D'].sum()
FN = round(FN, 2)
FN

151593.16

#### Answer: 15193.16

### Sklearn classification models are trained to maximize the accuracy. However, another error metric will be more relevant here. Which one? Please checkout make_scorer alongside with GridSearchCV in order to train the model to maximize the error metric of interest in this case.

In [19]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

max_depth_choices = [3,5,7]
min_samples_split_choices = [10, 20, 30]
min_samples_leaf_choices = [5, 10, 15, 20]
max_samples = [0.8]
random_state = [42]

param_grid = {
    'max_depth': max_depth_choices,
    'min_samples_split': min_samples_split_choices,
    'min_samples_leaf': min_samples_leaf_choices,
    'max_samples': max_samples,
    'random_state': random_state
}


model = RandomForestClassifier()
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5)
grid_search.fit(X1_train, y_train['TARGET_B'])

mean_test_scores = grid_search.cv_results_['mean_test_score']
params_list = grid_search.cv_results_['params']

grid_search.best_params_

{'max_depth': 7,
 'max_samples': 0.8,
 'min_samples_leaf': 5,
 'min_samples_split': 10,
 'random_state': 42}

In [20]:
for score, params in zip(mean_test_scores, params_list):
    print("Mean Test Score:", score)
    print("Hyperparameters:", params)
    print("-------------------")


Mean Test Score: 0.5913325512386999
Hyperparameters: {'max_depth': 3, 'max_samples': 0.8, 'min_samples_leaf': 5, 'min_samples_split': 10, 'random_state': 42}
-------------------
Mean Test Score: 0.5913325512386999
Hyperparameters: {'max_depth': 3, 'max_samples': 0.8, 'min_samples_leaf': 5, 'min_samples_split': 20, 'random_state': 42}
-------------------
Mean Test Score: 0.5913325512386999
Hyperparameters: {'max_depth': 3, 'max_samples': 0.8, 'min_samples_leaf': 5, 'min_samples_split': 30, 'random_state': 42}
-------------------
Mean Test Score: 0.5913325512386999
Hyperparameters: {'max_depth': 3, 'max_samples': 0.8, 'min_samples_leaf': 10, 'min_samples_split': 10, 'random_state': 42}
-------------------
Mean Test Score: 0.5913325512386999
Hyperparameters: {'max_depth': 3, 'max_samples': 0.8, 'min_samples_leaf': 10, 'min_samples_split': 20, 'random_state': 42}
-------------------
Mean Test Score: 0.5913325512386999
Hyperparameters: {'max_depth': 3, 'max_samples': 0.8, 'min_samples_leaf'

In [23]:
clf2 = RandomForestClassifier(max_depth=7,
                             min_samples_split=10,
                             min_samples_leaf =5,
                             max_samples=0.8,
                             random_state = 42)

clf2.fit(X1_train, y_train['TARGET_B'])

print("The Accuracy for the Random Forest in the TRAIN set is {:.2f}".format(clf2.score(X1_train, y_train['TARGET_B'])))
print("The Accuracy for the Random Forest in the TEST  set is {:.2f}".format(clf2.score(X1_test, y_test['TARGET_B'])))

y_pred2 = clf2.predict(X1_test)
display(y_test['TARGET_B'].value_counts())
display(confusion_matrix(y_test['TARGET_B'], y_pred2))

The Accuracy for the Random Forest in the TRAIN set is 0.68
The Accuracy for the Random Forest in the TEST  set is 0.67


1    18254
0    17974
Name: TARGET_B, dtype: int64

array([[11980,  5994],
       [ 6055, 12199]])

As we see, the numbers of False positives and False Negatives are reduced. The accurace rate increased by more than 0.06 in both tests. 